edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.IoTHub.Test/EdgeAgentConnectionTest.cs

```diff
@@ -749,6 +750,96 @@ public async Task GetDeploymentConfigInfoAsyncIncludesExceptionWhenDeserializeTh
             deviceManager.Verify(x => x.ReprovisionDeviceAsync(), Times.Exactly(shouldReprovision ? 1 : 0));
         }
 
+        [Fact]
+        [Unit]
+        public async Task FrequentTwinPullsOnConnectionAreThrottledAsync()
+        {
+            // Arrange
+            var deviceClient = new Mock<IModuleClient>();
+            deviceClient.Setup(x => x.UpstreamProtocol).Returns(UpstreamProtocol.Amqp);
+            deviceClient.Setup(x => x.IsActive).Returns(true);
+            var serde = new Mock<ISerde<DeploymentConfig>>();
+            ConnectionStatusChangesHandler connectionStatusChangesHandler = null;
+            var twin = new Twin
+            {
+                Properties = new TwinProperties
+                {
+                    Desired = new TwinCollection(
+                        JObject.FromObject(
+                            new Dictionary<string, object>
+                            {
+                                { "$version", 10 },
+                                { "MoreStuff", "MoreStuffHereToo" }
+                            }).ToString()),
+                    Reported = new TwinCollection()
+                }
+            };
+
+            var moduleClientProvider = new Mock<IModuleClientProvider>();
+            moduleClientProvider.Setup(d => d.Create(It.IsAny<ConnectionStatusChangesHandler>()))
+                .Callback<ConnectionStatusChangesHandler>(statusChanges => connectionStatusChangesHandler = statusChanges)
+                .ReturnsAsync(deviceClient.Object);
+
+            var counter = 0;
+            var milestone = new SemaphoreSlim(0, 1);
+
+            deviceClient.Setup(d => d.GetTwinAsync())
+                .ReturnsAsync(
+                    () =>
+                    {
+                        counter++;
+                        milestone.Release();
+
+                        return twin;
+                    });
+
+            serde.Setup(s => s.Deserialize(It.IsAny<string>())).Returns(() => DeploymentConfig.Empty);
+
+            IEnumerable<IRequestHandler> requestHandlers = new List<IRequestHandler> { new PingRequestHandler() };
+            var deviceManager = new Mock<IDeviceManager>();
+            deviceManager.Setup(x => x.ReprovisionDeviceAsync()).Returns(Task.CompletedTask);
+            Option<X509Certificate2> manifestTrustBundle = Option.None<X509Certificate2>();
+
+            var connection = new EdgeAgentConnection(moduleClientProvider.Object, serde.Object, new RequestManager(requestHandlers, DefaultRequestTimeout), deviceManager.Object, Mock.Of<IDeploymentMetrics>(), manifestTrustBundle);
+            connection.TwinPullOnConnectThrottleTime = TimeSpan.FromSeconds(3);
```

Comment:
minor: could we have smaller windows of less than a second? just to make this test take less than 10 seconds to complete?

Reply:
Yes, but I am a bit worried even with 3 sec that the test is not reliable. These are async events circulating in the system and we need some time to get those events. Probably it is way faster than 3 sec (I would guess it is in a 100ms magnitude) I can put it to 1.5 sec to see how it goes

Reply:
ok, I tried with 1,5sec then 2sec, but even with 2 sec it was not reliable, so I modified back to 3 sec